src/CodeGen_Hexagon.cpp

```diff
@@ -615,6 +615,23 @@ Value *CodeGen_Hexagon::interleave_vectors(const vector<llvm::Value *> &v) {
             }
             return concat_vectors(ret);
         }
+    } else if (v.size() == 3) {
+        // Interleaving 3 vectors - this generates awful code if we let LLVM do it,
+        // so we use vlut if we can. This is actually pretty general, it might be
+        // useful for other interleavings, though the LUT grows impractically
+        // large quite quickly.
+        if (element_bits == 8 || element_bits == 16) {
+            Value *lut = concat_vectors(v);
+
+            std::vector<int> indices;
+            for (unsigned i = 0; i < v_ty->getVectorNumElements(); i++) {
+                for (size_t j = 0; j < v.size(); j++) {
+                    indices.push_back(j * v_ty->getVectorNumElements() + i);
+                }
+            }
+
+            return vlut(lut, indices);
+        }
```

Comment:
I think if we warned on every bit of Halide IR that generates terrible code, we'd do nothing but add warnings for the next few days :)

In general, CodeGen_LLVM should decompose interleavings into 2 or 3 way interleavings, so the only really problematic ones are those with prime factors other than 2 or 3. If you do that in your code, this is probably the least of the problems.